heft-plugins/heft-typescript-plugin/src/TypeScriptPlugin.ts

```diff
@@ -0,0 +1,386 @@
+            '$.compilerOptions.outDir': {
+              pathResolutionMethod: PathResolutionMethod.resolvePathRelativeToConfigurationFile
+            }
+          }
+        });
+      }
+
+      partialTsconfigFilePromise = _partialTsconfigFileLoader.loadConfigurationFileForProjectAsync(
+        terminal,
+        buildFolder,
+        heftConfiguration.rigConfig
+      );
+    }
+    _partialTsconfigFilePromiseCache.set(buildFolder, partialTsconfigFilePromise);
+  }
+
+  return await partialTsconfigFilePromise;
+}
+
+export default class TypeScriptPlugin implements IHeftTaskPlugin {
+  public readonly pluginName: string = PLUGIN_NAME;
+
+  public accessor: ITypeScriptPluginAccessor = {
+    onChangedFilesHook: new SyncHook<IChangedFilesHookOptions>(['changedFilesHookOptions'])
+  };
+
+  public apply(taskSession: IHeftTaskSession, heftConfiguration: HeftConfiguration): void {
+    taskSession.hooks.clean.tapPromise(PLUGIN_NAME, async (cleanOptions: IHeftTaskCleanHookOptions) => {
+      await this._updateClean(taskSession, heftConfiguration, cleanOptions);
+    });
+
+    taskSession.hooks.run.tapPromise(PLUGIN_NAME, async (runOptions: IHeftTaskRunHookOptions) => {
+      await this._runTypeScriptAsync(taskSession, heftConfiguration);
+      await this._updateStaticAssetsToCopy(taskSession, heftConfiguration, runOptions);
+    });
+  }
+
+  private async _updateClean(
+    taskSession: IHeftTaskSession,
+    heftConfiguration: HeftConfiguration,
+    cleanOptions: IHeftTaskCleanHookOptions
+  ): Promise<void> {
+    const configurationFile: ITypeScriptConfigurationJson | undefined =
+      await loadTypeScriptConfigurationFileAsync(heftConfiguration, taskSession.logger.terminal);
+
+    // For now, delete the entire output folder and additional module kind output folders. In the future,
+    // we may want to clean specific files that we know are produced by the TypeScript compiler.
+    const tsconfigOutDir: string | undefined = await this._getTsconfigOutDirAsync(
+      taskSession,
+      heftConfiguration,
+      configurationFile
+    );
+    if (tsconfigOutDir) {
+      cleanOptions.addDeleteOperations({ sourceFolder: tsconfigOutDir });
+    }
+
+    if (configurationFile?.additionalModuleKindsToEmit) {
+      for (const additionalModuleKindToEmit of configurationFile.additionalModuleKindsToEmit) {
+        cleanOptions.addDeleteOperations({
+          sourceFolder: path.resolve(heftConfiguration.buildFolder, additionalModuleKindToEmit.outFolderName)
+        });
+      }
+    }
+  }
+
+  private async _updateStaticAssetsToCopy(
+    taskSession: IHeftTaskSession,
+    heftConfiguration: HeftConfiguration,
+    runOptions: IHeftTaskRunHookOptions
+  ): Promise<void> {
+    const typeScriptConfiguration: ITypeScriptConfigurationJson | undefined =
+      await loadTypeScriptConfigurationFileAsync(heftConfiguration, taskSession.logger.terminal);
+
+    // We only care about the copy if static assets were specified.
+    if (
```

Comment:
Can we move static asset copying out of the typescript plugin? It's a weird place to put it, even if I understand that it is there purely to accommodate having multiple output directories. Ideally we'd provide a standardized way to have static assets end up in a single output folder and have all the transpiled files point to that folder instead of cloning the files.

Reply:
In the spirit of making the upgrade less painful, I'm going to leave this as-is. Though that sounds like a good improvement for the future. I'll make a note in a comment.

Though, it's not there _just_ to accomodate multiple output directories. we do need to get the assets into an output folder, and it would be good if a developer didn't need to specify a manual copy in `heft.json` each time that this is done.